server/aesdsocket.c

```diff
@@ -61,12 +80,49 @@ int main(int argc, char *argv[])
 
     setup_daemon_mode();         /* Handle daemon mode if specified */
 
+    /* Create thread for appending timestamps every 10 seconds */
+    pthread_t timestamp_thread_id;
+    if (pthread_create(&timestamp_thread_id, NULL, timestamp_thread, NULL) != 0) {
+        syslog(LOG_ERR, "[ERROR] Failed to create timestamp thread.");
+        terminate_program(EXIT_FAILURE);
+    }
+
     /* Main server loop */
     while (!shutdown_flag)
     {
-        accept_client_and_handle_data(); /* Handle each client connection */
+        struct sockaddr_in client_address;
+        socklen_t client_address_len = sizeof(client_address);
+        int client_socket = accept(server_socket_fd, (struct sockaddr*)&client_address, &client_address_len);
+
+        if (client_socket == -1) {
+            syslog(LOG_WARNING, "[WARNING] Accepting connection failed, retrying...");
+            continue;
+        }
+
+        /* Spawn a new thread to handle the connection */
+        socket_thread_t *new_thread = (socket_thread_t *)malloc(sizeof(socket_thread_t));
+        if (new_thread == NULL) {
+            syslog(LOG_ERR, "[ERROR] Failed to allocate memory for new thread.");
+            close(client_socket);
+            continue;
+        }
+
+        new_thread->client_socket_fd = client_socket;
+        new_thread->thread_complete = false;
+
+        /* Add the new thread to the list */
+        SLIST_INSERT_HEAD(&thread_list_head, new_thread, entries);
+
+        if (pthread_create(&new_thread->thread_id, NULL, handle_client_connection, (void *)(intptr_t)client_socket) != 0) {
+            syslog(LOG_ERR, "[ERROR] Failed to create thread for client connection.");
+            close(client_socket);
+            free(new_thread);
+        }
     }
 
+    /* Join and clean up all threads before exiting */
+    join_and_cleanup_threads();
+
```

Comment:
Also note that you need a special behavior of `join_and_cleanup_threads` on shutdown which waits for threads to complete.  Otherwise you won't clean up the thread which hasn't exited yet.  This is what the valgrind warning is telling you.